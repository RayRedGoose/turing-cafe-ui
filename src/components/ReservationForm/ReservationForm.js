import React, { Component } from 'react'
import './ReservationForm.scss'
import Input from '../Input/Input'

class ReservationForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      name: '',
      date: '',
      time: '',
      number: ''
    }
  }

  updateEnteredData = (type, value) => {
    switch (type) {
      case 'name':
        this.setState({name: value})
        break;
      case 'date':
        this.setState({date: value})
        break;
      case 'time':
        this.setState({time: value})
        break;
      case 'number':
        this.setState({number: value})
        break;
      default:
        this.setState({error: 'Unvalid type'})
        break;
    }
  }

  submitReserv = (event) => {
    event.preventDefault()
    const newReserv = {...this.state, id: Date.now()}
    this.props.addNewReserv(newReserv)
    this.clearInputs()
  }

  clearInputs = () => {
    this.setState({
      name: '',
      date: '',
      time: '',
      number: ''
    })
  }

  render() {
    const inputs = Object.keys(this.state).map(key => {
      return (
        <Input
          key={key}
          type={key}
          currentValue={this.state[key]}
          action={this.updateEnteredData} />
      )
    })

    return (
      <form className="reserv-form">
        {inputs}
        <button onClick={this.submitReserv}>make reservation</button>
      </form>
    )
  }
}

export default ReservationForm
