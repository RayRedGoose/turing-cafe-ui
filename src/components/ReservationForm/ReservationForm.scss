.reserv-form {
  margin: 20px 40px;
  * {
    font-size: 16px;
    text-transform: capitalize;
  }
  button {
    background-color: #088081;
    color: #fff;
    border: none;
    padding: 0.4em 2em;
  }
}
