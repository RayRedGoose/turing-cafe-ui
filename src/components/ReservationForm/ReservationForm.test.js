import React from 'react'
import { shallow } from 'enzyme'
import ReservationForm from './ReservationForm'

describe("ReservationForm", () => {
  let form, instance
  const addNewReserv = jest.fn()

  beforeEach(() => {
    form = shallow(
      <ReservationForm
        addNewReserv={addNewReserv}/>
    )
    instance = form.instance()
  })

  it('renders with all data passed in correctlly', () => {
    expect(form).toMatchSnapshot()
  })

  describe("updateEnteredData", () => {
    it('should update name state', () => {
      expect(form.state('name')).toEqual('')

      instance.updateEnteredData('name', 'Mark')

      expect(form.state('name')).toEqual('Mark')
    })

    it('should update time state', () => {
      expect(form.state('time')).toEqual('')

      instance.updateEnteredData('time', '5:00')

      expect(form.state('time')).toEqual('5:00')
    })

    it('should update date state', () => {
      expect(form.state('date')).toEqual('')

      instance.updateEnteredData('date', '12/12')

      expect(form.state('date')).toEqual('12/12')
    })

    it('should update number state', () => {
      expect(form.state('number')).toEqual('')

      instance.updateEnteredData('number', '5')

      expect(form.state('number')).toEqual('5')
    })
  })

  describe("submitReserv", () => {
    it('should be called by button click', () => {
      const spy = jest.spyOn(instance, 'submitReserv').mockImplementation(() => {})
      instance.forceUpdate()

      form.find('button').simulate(
        'click',
        {preventDefault: jest.fn()}
      )

      expect(spy).toHaveBeenCalled()
    })

    it('should call addNewReserv prop', () => {
      instance.submitReserv({preventDefault: jest.fn()})

      expect(addNewReserv).toHaveBeenCalled()
    })

    it('should call clearInputs', () => {
      const spy = jest.spyOn(instance, 'clearInputs').mockImplementation(() => {})
      instance.forceUpdate()

      instance.submitReserv({preventDefault: jest.fn()})

      expect(spy).toHaveBeenCalled()
    })
  })
})
