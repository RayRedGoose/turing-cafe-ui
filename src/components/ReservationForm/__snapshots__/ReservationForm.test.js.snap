// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationForm renders with all data passed in correctlly 1`] = `
<form
  className="reserv-form"
>
  <Input
    action={[Function]}
    currentValue=""
    key="name"
    type="name"
  />
  <Input
    action={[Function]}
    currentValue=""
    key="date"
    type="date"
  />
  <Input
    action={[Function]}
    currentValue=""
    key="time"
    type="time"
  />
  <Input
    action={[Function]}
    currentValue=""
    key="number"
    type="number"
  />
  <button
    onClick={[Function]}
  >
    make reservation
  </button>
</form>
`;
